### PR TITLE
Conditionally serialize flushFrequency as int

### LINKED ISSUE #119325: failing kubelet / kubeadm skew tests
### Which jobs are failing?

https://testgrid.k8s.io/sig-cluster-lifecycle-kubeadm#kubeadm-kinder-kubelet-1-26-on-latest
https://testgrid.k8s.io/sig-cluster-lifecycle-kubeadm#kubeadm-kinder-kubelet-1-27-on-latest

also
https://testgrid.k8s.io/sig-cluster-lifecycle-kubeadm#kubeadm-kinder-latest-on-1-27

### Which tests are failing?

skewed / older kubelet is failing to parse a --config file generated by kubeadm at k/k master (1.28-pre).
kubeadm imports the public v1beta1 kubeletconfiguration, applies some kubeadm related defaults and then marshals the config to a file. since some of the kubelet fields are not pointers they end up as part of the marshaled output as per golang's PoD type defaults - e.g. bools are "false", ints are 0, etc.

```
kubeadm version
kubeadm version: &version.Info{Major:"1", Minor:"28+", GitVersion:"v1.28.0-alpha.4.212+e3bc35bc1bc7b3", GitCommit:"e3bc35bc1bc7b399b564f2c4efc75eb9959e70f7", GitTreeState:"clean", BuildDate:"2023-07-14T06:08:00Z", GoVersion:"go1.20.5", Compiler:"gc", Platform:"linux/amd64"}
```

```
kubelet --version
Kubernetes v1.27.3-50+3ca3e0ad4841ce
```

```
# modified minimal kubelet config to reproduce
#
cat /var/lib/kubelet/config.yaml
apiVersion: kubelet.config.k8s.io/v1beta1
kind: KubeletConfiguration
authentication:
  anonymous:
    enabled: false
  webhook:
    enabled: false
authorization:
  mode: AlwaysAllow
logging:
  flushFrequency: 0s # problem field
```

```
kubelet --config /var/lib/kubelet/config.yaml
E0714 11:21:44.209001    3335 run.go:74] "command failed" err="failed to load kubelet config file, error: failed to load Kubelet config file /var/lib/kubelet/config.yaml, error failed to decode: json: cannot unmarshal string into Go struct field LoggingConfiguration.logging.flushFrequency of type time.Duration, path: /var/lib/kubelet/config.yaml"
```

on a quick look it seems this old version of kubelet does not tolerate the "0s".
before https://github.com/kubernetes/kubernetes/pull/117800/commits/c550c17f7f0910ca32fcf97762170bea4f616cc6#diff-0eee074a0055cfbd121ea036feb1e593d7e37980cc291732368c0b94a8ded444L42-L45

```
logging:
  flushFrequency: 0
```

after:
 ```
logging:
  flushFrequency: 0s
```

which renders the change in the kubelet v1beta1 not backwards compatible for existing v1beta1 configs prepared by users.
we either need to allow the 0s to pass for older kubelets and/or backport some changes.

### Since when has it been failing?

since https://github.com/kubernetes/kubernetes/pull/117800 merged

cc @liggitt @pohly @danwinship 
/assign @cyclinder 

### Testgrid link

_No response_

### Reason for failure (if possible)

_No response_

### Anything else we need to know?

_No response_

### Relevant SIG(s)

/sig

### PR DESCRIPTION
**Pull request by liggitt**
#### What type of PR is this?

/kind bug

#### Which issue(s) this PR fixes:

Fixes #119325

This allows flushFrequency to serialize as an int or a string, defaults to serializing as an int for backwards compatibility for config constructors, round-trips faithfully when reading from a config file and writing back out.

#### Does this PR introduce a user-facing change?
```release-note
NONE
```

### CODE REVIEW

**Review by neolit123 — COMMENTED** [review]
/lgtm
/hold

thanks! please remove the hold if needed.

**Comment by pacoxu** [comment]
/hold
ci and test failed.
```
vendor/k8s.io/component-base/logs/api/v1/types_test.go:301:29: tt.tomd.String undefined (type *TimeOrMetaDuration has no field or method String)
vendor/k8s.io/component-base/logs/api/v1/types_test.go:302:92: tt.tomd.String undefined (type *TimeOrMetaDuration has no field or method String)
```

**Comment by cyclinder** [comment]
unit test should be updated.

```
			if err := tt.tomd.UnmarshalJSON(b); err == nil {
				if tt.wanted != tt.tomd.Duration.String() {
					t.Errorf("unexpected wanted for %s, wanted: %v, got: %v", tt.name, tt.wanted, tt.tomd.Duration.String())
				}
			} else {
				if err.Error() != tt.wanted {
					t.Errorf("UnmarshalJSON() error = %v", err)
				}
			}
			t.Log(tt.wanted,tt.tomd.Duration.String())
```

https://github.com/kubernetes/kubernetes/blob/13172cba5c0e1c6a076dbda4aeebbccaf658c7f1/staging/src/k8s.io/component-base/logs/api/v1/types_test.go#L301

**Comment by liggitt** [comment]
updated

File: pkg/kubelet/apis/config/v1beta1/defaults_test.go [hunk 1]
```diff
@@ -115,7 +116,7 @@ func TestSetDefaultsKubeletConfiguration(t *testing.T) {
 				VolumePluginDir:                           DefaultVolumePluginDir,
 				Logging: logsapi.LoggingConfiguration{
 					Format:         "text",
-					FlushFrequency: logsapi.TimeOrMetaDuration{Duration: 5 * time.Second},
+					FlushFrequency: logsapi.TimeOrMetaDuration{Duration: metav1.Duration{5 * time.Second}, SerializeAsString: true},
```
Comment (cyclinder) [inline pkg/kubelet/apis/config/v1beta1/defaults_test.go:119]:
```suggestion
                                        FlushFrequency: logsapi.TimeOrMetaDuration{Duration: metav1.Duration{Duration: 5 * time.Second}, SerializeAsString: true},

```

File: pkg/kubelet/apis/config/v1beta1/defaults_test.go [hunk 2]
```diff
@@ -239,7 +240,7 @@ func TestSetDefaultsKubeletConfiguration(t *testing.T) {
 				KernelMemcgNotification:     false,
 				Logging: logsapi.LoggingConfiguration{
 					Format:         "",
-					FlushFrequency: logsapi.TimeOrMetaDuration{Duration: 5 * time.Second},
+					FlushFrequency: logsapi.TimeOrMetaDuration{Duration: metav1.Duration{5 * time.Second}, SerializeAsString: true},
```
Comment (cyclinder) [inline pkg/kubelet/apis/config/v1beta1/defaults_test.go:243]:
```suggestion
                                        FlushFrequency: logsapi.TimeOrMetaDuration{Duration: metav1.Duration{Duration: 5 * time.Second}, SerializeAsString: true},

```

File: pkg/kubelet/apis/config/v1beta1/defaults_test.go [hunk 3]
```diff
@@ -339,7 +340,7 @@ func TestSetDefaultsKubeletConfiguration(t *testing.T) {
 				VolumePluginDir:        DefaultVolumePluginDir,
 				Logging: logsapi.LoggingConfiguration{
 					Format:         "text",
-					FlushFrequency: logsapi.TimeOrMetaDuration{Duration: 5 * time.Second},
+					FlushFrequency: logsapi.TimeOrMetaDuration{Duration: metav1.Duration{5 * time.Second}, SerializeAsString: true},
```
Comment (cyclinder) [inline pkg/kubelet/apis/config/v1beta1/defaults_test.go:343]:
```suggestion
					FlushFrequency: logsapi.TimeOrMetaDuration{Duration: metav1.Duration{Duration: 5 * time.Second}, SerializeAsString: true},
```

File: pkg/kubelet/apis/config/v1beta1/defaults_test.go [hunk 4]
```diff
@@ -481,7 +482,7 @@ func TestSetDefaultsKubeletConfiguration(t *testing.T) {
 				KernelMemcgNotification:     true,
 				Logging: logsapi.LoggingConfiguration{
 					Format:         "json",
-					FlushFrequency: logsapi.TimeOrMetaDuration{Duration: 5 * time.Second},
+					FlushFrequency: logsapi.TimeOrMetaDuration{Duration: metav1.Duration{5 * time.Second}, SerializeAsString: true},
```
Comment (cyclinder) [inline pkg/kubelet/apis/config/v1beta1/defaults_test.go:485]:
```suggestion
					FlushFrequency: logsapi.TimeOrMetaDuration{Duration: metav1.Duration{Duration: 5 * time.Second}, SerializeAsString: true},
```

File: pkg/kubelet/apis/config/v1beta1/defaults_test.go [hunk 5]
```diff
@@ -627,7 +628,7 @@ func TestSetDefaultsKubeletConfiguration(t *testing.T) {
...
```
Comment (cyclinder) [inline pkg/kubelet/apis/config/v1beta1/defaults_test.go:631]:
```suggestion
					FlushFrequency: logsapi.TimeOrMetaDuration{Duration: metav1.Duration{Duration: 5 * time.Second}, SerializeAsString: true},
```

File: pkg/kubelet/apis/config/v1beta1/defaults_test.go [hunk 6]
```diff
@@ -724,7 +725,7 @@ func TestSetDefaultsKubeletConfiguration(t *testing.T) {
 				VolumePluginDir:                           DefaultVolumePluginDir,
 				Logging: logsapi.LoggingConfiguration{
 					Format:         "text",
-					FlushFrequency: logsapi.TimeOrMetaDuration{Duration: 5 * time.Second},
+					FlushFrequency: logsapi.TimeOrMetaDuration{Duration: metav1.Duration{5 * time.Second}, SerializeAsString: true},
```
Comment (cyclinder) [inline pkg/kubelet/apis/config/v1beta1/defaults_test.go:728]:
```suggestion
					FlushFrequency: logsapi.TimeOrMetaDuration{Duration: metav1.Duration{Duration: 5 * time.Second}, SerializeAsString: true},
```

File: pkg/kubelet/apis/config/v1beta1/defaults_test.go [hunk 7]
```diff
@@ -813,7 +814,7 @@ func TestSetDefaultsKubeletConfiguration(t *testing.T) {
...
```
Comment (cyclinder) [inline pkg/kubelet/apis/config/v1beta1/defaults_test.go:817]:
```suggestion
					FlushFrequency: logsapi.TimeOrMetaDuration{Duration: metav1.Duration{Duration: 5 * time.Second}, SerializeAsString: true},
```

File: pkg/kubelet/apis/config/v1beta1/defaults_test.go [hunk 8]
```diff
@@ -902,7 +903,7 @@ func TestSetDefaultsKubeletConfiguration(t *testing.T) {
...
```
Comment (cyclinder) [inline pkg/kubelet/apis/config/v1beta1/defaults_test.go:906]:
```suggestion
					FlushFrequency: logsapi.TimeOrMetaDuration{Duration: metav1.Duration{Duration: 5 * time.Second}, SerializeAsString: true},
```

File: pkg/proxy/apis/config/v1alpha1/defaults_test.go [hunk 9]
```diff
@@ -70,7 +70,7 @@ func TestDefaultsKubeProxyConfiguration(t *testing.T) {
 				ConfigSyncPeriod: metav1.Duration{Duration: 15 * time.Minute},
 				Logging: logsapi.LoggingConfiguration{
 					Format:         "text",
-					FlushFrequency: logsapi.TimeOrMetaDuration{Duration: 5 * time.Second},
+					FlushFrequency: logsapi.TimeOrMetaDuration{Duration: metav1.Duration{5 * time.Second}, SerializeAsString: true},
```
Comment (cyclinder) [inline pkg/proxy/apis/config/v1alpha1/defaults_test.go:73]:
```suggestion
					FlushFrequency: logsapi.TimeOrMetaDuration{Duration: metav1.Duration{Duration: 5 * time.Second}, SerializeAsString: true},
```

File: pkg/proxy/apis/config/v1alpha1/defaults_test.go [hunk 10]
```diff
@@ -110,7 +110,7 @@ func TestDefaultsKubeProxyConfiguration(t *testing.T) {
...
```
Comment (cyclinder) [inline pkg/proxy/apis/config/v1alpha1/defaults_test.go:113]:
```suggestion
					FlushFrequency: logsapi.TimeOrMetaDuration{Duration: metav1.Duration{Duration: 5 * time.Second}, SerializeAsString: true},
```

**Review by cyclinder — COMMENTED** [review]
```
ERROR: cmd/kube-proxy/app/server_test.go:225:58: composites: k8s.io/apimachinery/pkg/apis/meta/v1.Duration struct literal uses unkeyed fields (govet)
ERROR: 				FlushFrequency: logsapi.TimeOrMetaDuration{Duration: metav1.Duration{5 * time.Second}, SerializeAsString: true},
```

**Comment by liggitt** [comment]
updated again

**Pull request merged**